### PR TITLE
Updated pull request for trustedqsl

### PR DESCRIPTION
**Pull request by k1mu**
### Please confirm your submission meets all the criteria

- [X] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [X] My pull request follows the instructions at [App Submission][submission].
- [X] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [X] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [X] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [X] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [X] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by k1mu** [comment]
Corrective actions: removed unnecessary cleanup actions, added required configuration (icons, screenshot). Changed wx build to cmake-ninja.

File: org.arrl.trustedqsl.json [hunk 1]
```diff
@@ -0,0 +1,148 @@
+{
+    "app-id" : "org.arrl.trustedqsl",
+    "runtime" : "org.freedesktop.Platform",
+    "runtime-version" : "22.08",
+    "sdk" : "org.freedesktop.Sdk",
+    "command" : "tqsl",
+    "finish-args" : [
+	"--share=network",
+	"--socket=x11",
```
Comment (hfiguiere) [inline org.arrl.trustedqsl.json:9]: `share=ipc` is  needed for X11.

File: org.arrl.trustedqsl.json [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	"--filesystem=home"
+     ],
+    "modules" : [
+        {
+            "name" : "openssl",
```
Comment (hfiguiere) [inline org.arrl.trustedqsl.json:14]: I will repeat what I said on the PR you closed: you do NOT need to build either openssl or curl.

File: org.arrl.trustedqsl.json [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+		"*.a",
+		"/share/man",
+		"/bin/mdb_copy",
+		"/bin/mdb_dump",
+		"/bin/mdb_load",
+		"/bin/mdb_stat"
+	    ]
+        },
+        {
+            "name" : "curl",
```
Comment (hfiguiere) [inline org.arrl.trustedqsl.json:71]: repeating once again: curl is in the runtime.

File: org.arrl.trustedqsl.json [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+	    "buildsystem" : "cmake-ninja",
+    	    "config-opts" : [
+		    "-DwxUSE_OPENGL=No",
+		    "-DwxUSE_HTML=Yes",
+		    "-DwxUSE_STREAMS=Yes",
+		    "-DwxUSE_WXHTML_HELP=Yes",
+		    "-DCMAKE_BUILD_TYPE=Release",
+		    "-DwxUSE_UNICODE=Yes"
+	    ],
+	    "sources" : [
```
Comment (hfiguiere) [inline org.arrl.trustedqsl.json:112]: sources is usually last in the module by good practice. This applies everywhere.

File: org.arrl.trustedqsl.json [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+			"/bin/wxrc",
+			"/bin/wxrc-3.2",
+			"/share/bakefile",
+			"/include",
+			"/lib/cmake"
+	    ]
+	},
+	{
+	    "name" : "tqsl",
+	    "buildsystem" : "cmake",
```
Comment (hfiguiere) [inline org.arrl.trustedqsl.json:131]: it doesn't build with cmake-ninja ?

File: org.arrl.trustedqsl.json [hunk 6]
```diff
@@ -0,0 +1,148 @@
+{
+    "app-id" : "org.arrl.trustedqsl",
+    "runtime" : "org.freedesktop.Platform",
+    "runtime-version" : "22.08",
+    "sdk" : "org.freedesktop.Sdk",
+    "command" : "tqsl",
+    "finish-args" : [
+	"--share=network",
+	"--socket=x11",
+	"--filesystem=home"
```
Comment (hfiguiere) [inline org.arrl.trustedqsl.json:10]: this is a broad filesystem permission.

File: org.arrl.trustedqsl.json (hunk 1 above)
Comment (k1mu) [reply org.arrl.trustedqsl.json:9]: Added - I missed this in the last round.

File: org.arrl.trustedqsl.json (hunk 2 above)
Comment (k1mu) [reply org.arrl.trustedqsl.json:14]: Well, duh. You're absolutely correct - a build without those works just fine. 
I'm removing this.

File: org.arrl.trustedqsl.json (hunk 3 above)
Comment (k1mu) [reply org.arrl.trustedqsl.json:71]: Removed.

File: org.arrl.trustedqsl.json (hunk 4 above)
Comment (k1mu) [reply org.arrl.trustedqsl.json:112]: Corrected - moved the sources blocks to the end.

File: org.arrl.trustedqsl.json (hunk 5 above)
Comment (k1mu) [reply org.arrl.trustedqsl.json:131]: Yes, it does. Changed.

File: org.arrl.trustedqsl.json (hunk 6 above)
Comment (k1mu) [reply org.arrl.trustedqsl.json:10]: TrustedQSL stores configuration data in $HOME/.tqsl and can be used to sign log files from other logging applications.  I don't think full filesystem access is needed, but the way this is used (primarily as a utility for loggers to sign and upload), home access is needed.

**Comment by k1mu** [comment]
I've pushed an update based on the comments so far.

File: org.arrl.trustedqsl.json (hunk 6 above)
Comment (hfiguiere) [reply org.arrl.trustedqsl.json:11]: then you can use `--persist` instead.

ideally it would use xdg standard locations.

**Comment by hfiguiere** [comment]
bot, build org.arrl.trustedqsl

**Comment by hfiguiere** [comment]
tha app-id in the appstream file is incorrect.

File: org.arrl.trustedqsl.json (hunk 6 above)
Comment (k1mu) [reply org.arrl.trustedqsl.json:11]: I'll try with --persist.

**Comment by k1mu** [comment]
I see the error, will correct and resubmit.

**Comment by k1mu** [comment]
Trying with --persist versus --filesystem=home effectively makes this unusable. None of the configuration data (certificates, locations, etc.) is visible from the home folder and there's no way to import any of the necessary data. Opening files from outside of the container also fails, which is the major purpose of the app - to take a file exported by another application, read that for contact data, sign it, then save or upload the resulting container. 

I understand the desire to restrict what data applications have access to, but this can't run with container-only access.

File: org.arrl.trustedqsl.json (hunk 6 above)
Comment (AsciiWolf) [reply org.arrl.trustedqsl.json:11]: `--persist=.tqsl` should work fine.

**Comment by k1mu** [comment]
I've tried a clean build to verify the behavior with --persist=.tqsl and that's not going to work.
Let me explain this application and what it does, with hopes of making the case for --filesystem=home.

This is an application that has two facets - a GUI tool and a command line tool.
The GUI manages a PKI for the user - generating a certificate request and either saving it for later upload, or directly uploading that request.  The central system performs an identity verification step (which can include using the mail to validate that the requester is indeed the license holder) and issues a signed PKCS6 file back to the user once they're verified. There's some unique certificate extensions to indicate what "country" someone is  from (the country that issued their ham license) as well as a separate set of validity dates. (The certificate might expire in 3 years, but that ham license has a separate validity period that could be just a small number of days.) This is why there's a specialized PKI.
In addition, the GUI allows managing the data about the physical location of the operator - State, County, etc.

That gui management is relatively infrequent, but the data needs to be persistent and portable between computer systems. There's a backup/restore mechanism to allow populating a new computer with the operator's details.

Then, given that setup, a user submits a file of radio contacts in a standardized form (ADIF - Amateur Data Interchange Format).  those contacts are augmented with a digital signature verifying that the operator made that radio contact. The resulting signed file is either saved for later upload, or directly uploaded if there's an Internet connection.

The centralized server, Logbook of the World, matches contacts - station "A" says they contacted station "B" at a given date, time, and other details. Matches are generated and can be used for awards - operating a certain number of countries.

The majority of the time, radio contacts are managed by completely separate logging programs. Those emit ADIF-format logs and use the application's command line interface to digitally sign and optionally upload contacts. It's fairly rare for the users to manually upload themselves (but there's the ability to do that, as well as a GUI menu for creating and editing ADIF files.)

All of that means that an arbitrary program has to be able to pass "tqsl random-file.adif" and have it read, process, and upload the result. So TQSL has to be able to open files wherever that application put them.  I've tried with the mentioned build with "-persist=.tqsl" and I'm not able to open any files outside of that directory.  In addition, the existing ".tqsl" folder isn't being mapped, so  the existing configuration is lost. I can't even copy a backup package to that folder as it's not visible outside of the container.

This application is ancient - initially started in 2001.

tl;dr - think of this as a filter that takes input files and transforms them. I've got to be able to get input from almost anywhere; While I can state that there's a limitation for the flatpak build that any input files have to be somewhere under $HOME I can't expect several logging programs to adapt to a highly restrictive folder structure.

That all being said, I've been wildly off base already here. I am not trying to be argumentative and would be happy to discuss or be proven wrong yet again.

**Comment by hfiguiere** [comment]
> Let me explain this application and what it does, with hopes of making the case for --filesystem=home.

Then `--filesystem=~/.tqsl`

**Comment by k1mu** [comment]
> > Let me explain this application and what it does, with hopes of making the case for --filesystem=home.
> 
> Then `--filesystem=~/.tqsl`

That does get me to the existing configuration data, so that's a start.  However, wxWidgets stores config data in a dotfile off of home, so "--filesystem=~/.tqslapp" is needed as well. But that starts a slippery slope. 

Then how is the application supposed to open file ~/cqrlog/k1mu.adif when that application saves the file?
If the user saves a certificate as ~/callsign.tq6, how can they import it?

I can't dictate this big of a change on the dependent programs. And  having this be so restrictive on Linux versus MacOS/BSD/Windows would be a big downgrade - as well as being a significant regression from the packages that some distributions already have in place (many of which aren't being maintained, thus the idea of a Flatpak build so I can manage this.)
-

**Comment by k1mu** [comment]
Another option here that might be worth trying is to add the xdg-documents and xdg-downloads folders. I haven't tried that to see how it works, but that may be worth considering.

**Comment by hfiguiere** [comment]
AppStream fail to validate.
```
builddir/files/share/appdata/org.arrl.trustedqsl.appdata.xml: FAILED:
• tag-missing           : <release> required
Validation of files failed
```

**Comment by hfiguiere** [comment]
bot, build org.arrl.trustedqsl

**Comment by hfiguiere** [comment]
the icon is not found.

**Comment by hfiguiere** [comment]
```
-- Installing: /app/share/app-info/hicolor/icons/128x128/org.arrl.trustedqsl.png
-- Installing: /app/share/app-info/hicolor/icons/64x64/org.arrl.trustedqsl.png
-- Installing: /app/share/app-info/hicolor/icons/48x48/org.arrl.trustedqsl.png
-- Installing: /app/share/app-info/hicolor/icons/32x32/org.arrl.trustedqsl.png
-- Installing: /app/share/app-info/hicolor/icons/16x16/org.arrl.trustedqsl.png
```

This is  very wrong as icon path.

**Comment by barthalion** [comment]
bot, build org.arrl.trustedqsl

**Comment by barthalion** [comment]
Failed `Check for 128x128 icon` typically means a mismatch somewhere between the appdata, desktop and icon files. The compose command first reads the appdata file, uses ID from it (alternatively, refers to `<launchable>`) to look up the desktop file, from which it reads the `Icon` field. Everything has to match the app ID used in the manifest.

**Comment by hfiguiere** [comment]
I don't see a .desktop file installed.

**Comment by k1mu** [comment]
Thanks - that's easily fixed.
I've run the desktop validator against the desktop file, removed the icon spec from the metadata XML file, and added to the desktop. The icon is now being installed to the "right" directory.  Will update the package once I check the build output.

**Comment by hfiguiere** [comment]
bot, build org.arrl.trustedqsl

**Comment by barthalion** [comment]
It segfaults on my end:

```
[bpiotrowski@purei ~]$ coredumpctl info
           PID: 37484 (tqsl)
           UID: 1000 (bpiotrowski)
           GID: 1000 (bpiotrowski)
        Signal: 11 (SEGV)
     Timestamp: Fri 2023-05-12 16:14:51 CEST (30s ago)
  Command Line: tqsl
    Executable: /app/bin/tqsl
 Control Group: /user.slice/user-1000.slice/user@1000.service/app.slice/app-flatpak-org.arrl.trustedqsl-37467.scope
          Unit: user@1000.service
     User Unit: app-flatpak-org.arrl.trustedqsl-37467.scope
         Slice: user-1000.slice
     Owner UID: 1000 (bpiotrowski)
       Boot ID: c75385931ced49a58ed473a5a22c2a09
    Machine ID: b245857f888a412b840acce4c4cc3577
      Hostname: purei
       Storage: /var/lib/systemd/coredump/core.tqsl.1000.c75385931ced49a58ed473a5a22c2a09.37484.1683900891000000.zst (present)
  Size on Disk: 1.4M
       Message: Process 37484 (tqsl) of user 1000 dumped core.
                
                Stack trace of thread 2:
                #0  0x00007f9a57e7bc35 n/a (/usr/lib/x86_64-linux-gnu/libc.so.6 + 0x7bc35)
                #1  0x0000555958df95c0 n/a (/app/bin/tqsl + 0x14f5c0)
                #2  0x00007f9a58790538 n/a (/app/lib/libwx_baseu-3.2.so.0.2.1 + 0x190538)
                #3  0x0000555958d8ad34 n/a (/app/bin/tqsl + 0xe0d34)
                #4  0x00007f9a57e2954a n/a (/usr/lib/x86_64-linux-gnu/libc.so.6 + 0x2954a)
                #5  0x00007f9a57e2960b n/a (/usr/lib/x86_64-linux-gnu/libc.so.6 + 0x2960b)
                #6  0x0000555958d8fb75 n/a (/app/bin/tqsl + 0xe5b75)
                ELF object binary architecture: AMD x86-64
```

**Comment by k1mu** [comment]
The segfault is because there was insufficient checking of the code that creates the $HOME/.tqsl folder - when unable to do that, it'd eventually cause a null pointer dereference. A fix has been committed to the repo, will be pushing an update shortly.

**Comment by barthalion** [comment]
bot, build org.arrl.trustedqsl
